p2p/server.go

```diff
@@ -571,16 +573,34 @@ func (srv *Server) setupDiscovery() error {
 	}
 	realaddr := conn.LocalAddr().(*net.UDPAddr)
 	srv.log.Debug("UDP listener up", "addr", realaddr)
+
+	intport, extport := realaddr.Port, realaddr.Port
 	if srv.NAT != nil {
 		if !realaddr.IP.IsLoopback() {
+			srv.log.Info("Start Adding Maps to NAT")
+			// [NEED TO REMOVE] In this case, if the NAT doesn't respond, it will block for
+			// a long time. Have to decide whether we want to do the rest after the mapping,
+			// or run the goroutine and do the work first.
+			//
+			// Currently, it is not possible to distinguish whether the log is about for
+			// 'ethereum p2p' or 'ethereum discovery'.
+			p, err := srv.NAT.AddMapping("udp", intport, extport, "", nat.DefaultMapTimeout)
```

Comment:
Not sure why you're adding this synchronous call here. It's fine to change the port number assigned by NAT from the mapping goroutine, i.e. it's ok to call `SetFallbackUDP` in the goroutine.

Reply:
Thank you. Done!